### PR TITLE
add problem model, can create new problem

### PR DESCRIPTION
**Pull request by geniusgordon**

### CODE REVIEW

File: problem/forms.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+from django import forms
```
Comment (nickelcheng) [inline problem/forms.py]: Please add license.

**Comment by nickelcheng** [comment]
You should merge latest code before starting coding QQ

**Comment by geniusgordon** [comment]
啊那怎麼辦0.0

**Comment by nickelcheng** [comment]
You can merge it now and fix all conflicts.

File: problem/templates/problem/edit.html [hunk 2]
```diff
@@ -99,8 +99,7 @@
 $('#preview_button').click(function() {
   MyWindow = window.open('{% url "problem.views.preview" %}',
     'MyWindow',
-    'toolbar=no,location=no,directories=no,status=no,menubar=no,
-    scrollbars=yes,resizable=yes,width=600,height=30'
+    'toolbar=no,location=no,directories=no,status=no,menubar=no,scrollbars=yes,resizable=yes,width=600,height=30'
```
Comment (nickelcheng) [inline problem/templates/problem/edit.html]: Please take care of the indent.

File: problem/templates/problem/problem.html [hunk 3]
```diff
@@ -25,7 +25,7 @@
 <div style='height:90px;' class='panel panel-default'>
   <div style='padding-left:10px;padding-right:10px'>
     <div style='float:left;width:60%;'>
-      <h2>{{ title }} &nbsp;&nbsp;<a href="{% url 'problem.views.edit' pid%}">EDIT</a></h2>
+      <h2>{{ problem.name }} &nbsp;&nbsp;<a href="{% url 'problem.views.edit' problem.pk %}">EDIT</a></h2>
```
Comment (nickelcheng) [inline problem/templates/problem/problem.html]: I suggest spiting this line into two lines.

File: problem/views.py [hunk 4]
```diff
@@ -30,24 +32,29 @@ def problem(request):
...
-      'description': 'Given a, b, output a+b.',
-      'input': 'a, b <= 100000000',
-      'output': 'a+b',
-      'samp_input': '1 2\n4 5',
-      'samp_output': '3\n9\n',
-      'tag': [''],
-      'testcase': [{'num': 1, 'time': 1, 'memory': 32}, {'num': 2, 'time': 3, 'memory': 100}],
-    }
-    return render(request, 'problem/detail.html', p)
+    print 'problem', problem_id
```
Comment (nickelcheng) [inline problem/views.py]: Why do you print here?

**Comment by CHIAHSINHOU** [comment]
Do you make any buttons to link the create problem page in index page ?

**Comment by CHIAHSINHOU** [comment]
There is something wrong with checkbox.

**Comment by CHIAHSINHOU** [comment]
文字跟下面的格子沒有對齊 有點擠

File: problem/views.py [hunk 5]
```diff
@@ -30,24 +41,86 @@ def problem(request):
...
+                   'input': problem.input, 'output': problem.output,
+                   'sample_in': problem.sample_in, 'sample_out': problem.sample_out,
+                   'testcase': testcase })
 
 def new(request):
-    return render(request, 'problem/edit.html')
+    if request.method == 'GET':
+        form = ProblemForm()
+    if request.method == 'POST':
+        logger.info('post new problem')
```
Comment (nickelcheng) [inline problem/views.py]: Please add pid info.

File: problem/views.py [hunk 6]
```diff
@@ -30,24 +41,86 @@ def problem(request):
...
+            problem.sample_out = request.POST['sample_output']
+            problem.save()
+            return redirect('/problem/%d' % (problem.pk))
+    return render(request, 'problem/edit.html', { 'form': form, 'is_new': True })
+
+def tag(request, problem_id):
+    if request.method == 'POST':
+        tag = request.POST['tag']
+        problem = Problem.objects.get(pk=problem_id)
+        print problem.tags.filter(tag_name=tag).exists()
```
Comment (nickelcheng) [inline problem/views.py]: Don't use print directly. You may return an error message and show it on web page.

File: problem/views.py [hunk 7]
```diff
@@ -30,24 +41,86 @@ def problem(request):
...
-      'description': 'Given a, b, output a+b.',
-      'input': 'a, b <= 100000000',
-      'output': 'a+b',
-      'samp_input': '1 2\n4 5',
-      'samp_output': '3\n9\n',
-      'tag': [''],
-      'testcase': [{'num': 1, 'time': 1, 'memory': 32}, {'num': 2, 'time': 3, 'memory': 100}],
-    }
-    return render(request, 'problem/detail.html', p)
+    logger.info('detail of problem %s' % (problem_id))
```
Comment (nickelcheng) [inline problem/views.py]: query db does not need to log.

File: problem/views.py [hunk 8]
```diff
@@ -30,24 +41,90 @@ def problem(request):
...
-      'description': 'Given a, b, output a+b.',
-      'input': 'a, b <= 100000000',
-      'output': 'a+b',
-      'samp_input': '1 2\n4 5',
-      'samp_output': '3\n9\n',
-      'tag': [''],
-      'testcase': [{'num': 1, 'time': 1, 'memory': 32}, {'num': 2, 'time': 3, 'memory': 100}],
-    }
-    return render(request, 'problem/detail.html', p)
+    problem = Problem.objects.get(pk=problem_id)
```
Comment (nickelcheng) [inline problem/views.py]: You should check whether the given problem id is valid or not.
If the problem does not exist, write a warning log.

File: problem/views.py [hunk 9]
```diff
@@ -30,24 +41,90 @@ def problem(request):
...
+    testcase = Testcase.objects.filter(problem=problem)
+    tag = problem.tags.all()
+    return render(request, 'problem/detail.html', 
+                  { 'problem': problem, 'tag': tag, 'testcase': testcase })
 
 def edit(request, problem_id):
-    return render(request, 'problem/edit.html')
+    logger.info('edit problem %s' % (problem_id))
```
Comment (nickelcheng) [inline problem/views.py]: I prefer logging after users submit their modifications.

File: problem/views.py [hunk 10]
```diff
@@ -30,24 +41,90 @@ def problem(request):
...
+            problem.tags.add(new_tag)
+            problem.save()
+            return HttpResponse()
+        return HttpResponseBadRequest()
+    return HttpResponse()
+
+def testcase(request, problem_id, tid=None):
+    if request.method == 'POST':
+        if tid == None:
+            logger.info('new test case')
```
Comment (nickelcheng) [inline problem/views.py]: I prefer adding tid in log for new testcase.

File: problem/views.py [hunk 11]
```diff
@@ -30,24 +41,90 @@ def problem(request):
...
+            testcase = Testcase()
+        else:
+            logger.info('update test case, tid = %s' % (tid))
+            testcase = Testcase.objects.get(pk=tid)
```
Comment (nickelcheng) [inline problem/views.py]: I think it's better to check whether the id is valid or not.

File: problem/views.py [hunk 12]
```diff
@@ -30,24 +41,90 @@ def problem(request):
...
+        testcase.problem = Problem.objects.get(pk=problem_id)
```
Comment (nickelcheng) [inline problem/views.py]: Ditto.

**Pull request closed without merging**